hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeeder.java

```diff
@@ -134,27 +141,75 @@ private void handleUncaughtThrowable(Throwable t) {
     }
 
     private boolean fillBuffer() {
-        // iterate over inputs and execute default target
+        Bundle p = nextSourceBundle();
+        if (p == null) {
+            return false;
+        }
+        int hash = p.hashCode();
+        if (shardField != null) {
+            String val = ValueUtil.asNativeString(p.getValue(p.getFormat().getField(shardField)));
+            if (!Strings.isNullOrEmpty(val)) {
+                hash = PluggableHashFunction.hash(val);
+            }
+        }
+        int mod = Math.abs(hash % queues.length);
         try {
-            Bundle p = source.next();
-            if (p == null) {
-                log.info("exiting on null bundle from {}", source);
+            queues[mod].put(p);
+        } catch (InterruptedException e) {
+            if (!batchFillAllDrainedQueues(mod, p)) {
                 return false;
             }
-            int hash = p.hashCode();
-            if (shardField != null) {
-                String val = ValueUtil.asNativeString(p.getValue(shardField));
-                if (!LessStrings.isEmpty(val)) {
-                    hash = PluggableHashFunction.hash(val);
+        }
+        return true;
+    }
+
+    private Bundle nextSourceBundle() {
+        Bundle bundle = null;
+        try {
+            bundle = source.next();
+        } catch (NoSuchElementException e) {
+        }
+        if (bundle == null) {
+            log.info("exiting on null bundle or NoSuchElementException from {}", source);
+        }
+        return bundle;
+    }
+
+    private boolean batchFillAllDrainedQueues(int interruptedQueue, @Nonnull Bundle interruptedBundle) {
+        boolean interruptedBundlePutBack = false;
+        int fillCount;
+        for (MapperTask t : mapperTasks) {
+            if (t.queueDrained) {
+                t.queueDrained = false;
+                if (!interruptedBundlePutBack) {
+                    pushQueue(t.processorID, interruptedBundle);
+                    interruptedBundlePutBack = true;
+                    fillCount = QUEUE_DEPTH - 1;
+                } else {
+                    fillCount = QUEUE_DEPTH;
+                }
+                if (!fillQueue(t.processorID, fillCount)) {
+                    return false;
                 }
             }
-            int mod = Math.abs(hash % queues.length);
-            pushQueue(mod, p);
-            return true;
-        } catch (NoSuchElementException ignored) {
-            log.info("exiting on premature stream termination");
         }
-        return false;
+        if (!interruptedBundlePutBack) {
+            log.error("Batch fill is called when no queue is drained. This is a bug!");
```

Comment:
If this error will lead to data being processed incorrectly then we should throw an IllegalStateException and have the job error.